File: problems/dp/lc_416-分割等和子集.py
from typing import List

"""

转化为0-1背包问题

背包大小为 sum(nums) // 2

dp[i][j] 背包大小为 i, 前 j 个物品

"""
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        total = sum(nums)
        if total % 2 != 0:
            return False
        
        capacity = total // 2
        n = len(nums)
        dp = [[0] * (n + 1) for _ in range(capacity + 1)]

        for i in range(1, capacity + 1):
            for j in range(1, n + 1):
                if nums[j - 1] <= i:
                    dp[i][j] = max(dp[i][j - 1], dp[i - nums[j - 1]][j - 1] + nums[j - 1])
                else:
                    dp[i][j] = dp[i][j - 1]

        return dp[capacity][n] == capacity


print(Solution().canPartition([2, 1, 1, 2]))
        
"""

创建二维数组 dp，包含 target+1 行与 n+1 列，其中 dp[i][j] 表示从数组的前 j 个数内选取若干个正整数（可以是 0 个），
是否存在一种选取方案使得被选取的正整数的和等于 i。初始时，dp 中的全部元素都是 false。

"""
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        n = len(nums)
        if n < 2:
            return False
        
        total = sum(nums)
        if total % 2 != 0:
            return False
        
        target = total // 2
        dp = [[False] * (n + 1) for _ in range(target + 1)]
        
        dp[0][0] = True

        for i in range(1, target + 1):
            for j in range(1, n + 1):
                if nums[j - 1] <= i:
                    dp[i][j] = dp[i][j - 1] | dp[i - nums[j - 1]][j - 1]
                else:
                    dp[i][j] = dp[i][j - 1]

        return dp[target][n]

print(Solution().canPartition([2, 1, 1, 2]))


"""

还可以继续优化，使用一维数组进行优化空间

nums 对应于行
target 对应于列

"""

class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        n = len(nums)
        if n < 2:
            return False
        
        total = sum(nums)
        if total % 2 != 0:
            return False
        
        target = total // 2
        dp = [True] + [False] * target

        for i, x in enumerate(nums):
            for j in range(target, x - 1, -1):
                dp[j] = dp[j] | dp[j - x]

        return dp[target]

print(Solution().canPartition([2, 1, 1, 2]))

File: problems/dp/lc_416-分割等和子集.cpp
#include <algorithm>
#include <bitset>
#include <iostream>
#include <numeric>
#include <vector>

namespace dp1 {

class Solution
{
public:
  bool canPartition(const std::vector<int> &nums)
  {
    int n = nums.size();
    if (n < 2) {
      return false;
    }

    int sum = std::accumulate(nums.begin(), nums.end(), 0);

    if ((sum & 0x1) != 0) {
      return false;
    }

    int target = sum >> 1;

    // 物品价值与重量相等
    // dp[i][j] 表示背包大小为 j 时, 装前 i 个物品得到的最大价值
    std::vector<std::vector<int>> dp(nums.size() + 1, std::vector<int>(target + 1, 0));
    for (int i = 1; i <= n; ++i) {
      for (int j = 1; j <= target; ++j) {  // j 为背包大小
        if (j >= nums[i - 1]) {
          dp[i][j] = std::max(dp[i - 1][j], dp[i - 1][j - nums[i - 1]] + nums[i - 1]);
        } else {
          dp[i][j] = dp[i - 1][j];
        }
      }
    }

    return dp[n][target] == target;  // 正好能装满一半
  }
};

}  // namespace dp1

namespace dp2 {

class Solution
{
public:
  bool canPartition(const std::vector<int> &nums)
  {
    int n = nums.size();
    if (n < 2) {
      return false;
    }

    int sum = std::accumulate(nums.begin(), nums.end(), 0);

    // 不是偶数
    if ((sum & 0x1) != 0) {
      return false;
    }

    int target = sum >> 1;

    // 物品价值与重量相等
    // dp[i][j] 表示背包大小为 j 时, 装前 i 个物品得到的最大价值
    std::vector<std::vector<bool>> dp(nums.size() + 1, std::vector<bool>(target + 1, false));
    dp[0][0] = true;
    for (int i = 1; i <= n; ++i) {
      for (int j = 1; j <= target; ++j) {  // j 为背包大小
        if (j >= nums[i - 1]) {
          dp[i][j] = dp[i - 1][j] || dp[i - 1][j - nums[i - 1]];
        } else {
          dp[i][j] = dp[i - 1][j];
        }
      }
    }

    return dp[n][target];
  }
};

}  // namespace dp2

namespace dp3 {

class Solution
{
public:
  bool canPartition(const std::vector<int> &nums)
  {
    int n = nums.size();
    if (n < 2) {
      return false;
    }

    int sum = std::accumulate(nums.begin(), nums.end(), 0);

    // 不是偶数
    if ((sum & 0x1) != 0) {
      return false;
    }

    int target = sum >> 1;

    std::vector<bool> dp(target + 1, false);
    dp[0] = true;

    for (auto num : nums) {
      for (int j = target; j >= num; --j) {
        dp[j] = dp[j] || dp[j - num];
      }
    }
    return dp[target];
  }
};

}  // namespace dp3

namespace dp4 {

class Solution
{
public:
  bool canPartition(const std::vector<int> &nums)
  {
    //why 10001? cover all the cases of sum of subarr 100 * 200 / 2 = 10000
    std::bitset<10001> f;
    int sum = 0;
    f[0]    = true;
    for (int num : nums) {
      f    = f << num | f; // 直接消去一个 for 循环
      sum += num;
    }
    if ((sum & 1) != 0) {
      return false;
    }
    return f[sum >> 1];
  }
};

}  // namespace dp4


int main()
{
  std::cout << dp1::Solution().canPartition({1, 5, 11, 5}) << '\n';
  std::cout << dp2::Solution().canPartition({1, 5, 11, 5}) << '\n';
  std::cout << dp3::Solution().canPartition({1, 5, 11, 5}) << '\n';
  std::cout << dp4::Solution().canPartition({1, 5, 11, 5}) << '\n';
}